### PR TITLE
Version locked items can't be moved or changed

### PR DESCRIPTION
**Pull request by jonathan-s**
Also refactors urs so that the admin can be used with other models that implements the same interface. 

Superseeded by https://github.com/FidelityInternational/djangocms-navigation/pull/56

### CODE REVIEW

**Comment by jonathan-s** [comment]
Requires changes to some other files as well. So will be on that.

File: tests/test_admin.py [hunk 1]
```diff
@@ -131,12 +132,63 @@ def test_get_queryset_filters_by_content_id(self):
...
+            "admin:djangocms_navigation_menuitem_move_node", args=(menu_content.id,)
+        )
+        self.data = {
+            "node_id": child_of_child.pk,
+            "sibling_id": menu_content.root.pk,
+            "as_child": 1,
+        }
+
+    def test_visit_change_view_when_node_is_version_locked_fails(self):
+        """It fails as super user is not the person who created the version"""
```
Comment (Aiky30) [inline tests/test_admin.py]: The docstrings should be on every test and should describes the business requirement with what and why. In this instance this could be: Visiting the change view for a menu item by a different user should be rejected, because the user is different to the locked to user.

This helps anyone making changes in the future who could break this test understand what it was trying to do, if it then needs to be reworked it can be as the developer should be informed of the requirement that the test is enforcing.

File: tests/test_admin.py [hunk 2]
```diff
@@ -131,12 +132,63 @@ def test_get_queryset_filters_by_content_id(self):
...
+        response = self.client.get(self.change_url)
+        msg = list(get_messages(response.wsgi_request))[0]
+
+        self.assertEquals(msg.message, "You don't have permission to edit or it is locked")
+        self.assertEquals(response.status_code, 302)
+
+    def test_moving_node_that_is_version_locked_fails(self):
```
Comment (Aiky30) [inline tests/test_admin.py]: A menu item cannot be changed by a user who is not the lock owner.

File: tests/test_admin.py [hunk 3]
```diff
@@ -131,12 +132,63 @@ def test_get_queryset_filters_by_content_id(self):
...
+        msg = "The item is currently locked or you don't have permission to change it"
+
+        self.assertEquals(response.status_code, 400)
+        self.assertEquals(content, msg)
+
+    @patch('djangocms_navigation.admin.using_version_lock', False)
+    def test_moving_node_version_lock_not_installed_works_without_error(self):
+        response = self.client.post(self.move_url, data=self.data)
+
+        self.assertEquals(response.status_code, 200)
```
Comment (Aiky30) [inline tests/test_admin.py]: We should really check that the page is that expected rather than just a 200 here.

File: tests/test_admin.py [hunk 4]
```diff
@@ -131,12 +132,63 @@ def test_get_queryset_filters_by_content_id(self):
...
+    def test_moving_node_that_is_version_locked_fails(self):
+        response = self.client.post(self.move_url, data=self.data)
+        content = response.content.decode('utf-8')
+        msg = "The item is currently locked or you don't have permission to change it"
+
+        self.assertEquals(response.status_code, 400)
+        self.assertEquals(content, msg)
+
+    @patch('djangocms_navigation.admin.using_version_lock', False)
+    def test_moving_node_version_lock_not_installed_works_without_error(self):
```
Comment (Aiky30) [inline tests/test_admin.py]: """A menu item can be moved without issue if version locking is not installed
"""

File: tests/test_admin.py [hunk 5]
```diff
@@ -131,12 +132,63 @@ def test_get_queryset_filters_by_content_id(self):
...
+        response = self.client.post(self.move_url, data=self.data)
+
+        self.assertEquals(response.status_code, 200)
+
+
 class MenuItemAdminChangeViewTestCase(CMSTestCase, UsefulAssertsMixin):
     def setUp(self):
         self.client.force_login(self.get_superuser())
 
     def test_menuitem_change_view(self):
```
Comment (Aiky30) [inline tests/test_admin.py]: We should fix this and describe what's being tested in this test. 

"""The change view does not show draft versions"""

This test is a perfect example of why the docstring should always be in a test, I had to work out what the test was doing by reading code. That should not be the case, imagine if 10's of tests were failing and you had to work out if the test needed reworking or the code that it checks. If you know "why" something is being tested then you can decide easily which one to change.

File: tests/test_admin.py [hunk 6]
```diff
@@ -592,10 +646,11 @@ def test_menuitem_changelist_redirects_if_not_latest_version(self, mocked_messag
 
 class MenuItemAdminMoveNodeViewTestCase(CMSTestCase):
     def setUp(self):
-        self.client.force_login(self.get_superuser())
+        self.user = self.get_superuser()
+        self.client.force_login(self.user)
 
     def test_menuitem_move_node_smoke_test(self):
```
Comment (Aiky30) [inline tests/test_admin.py]: """A node can be moved"""

File: tests/test_admin.py [hunk 7]
```diff
@@ -616,7 +671,7 @@ def test_menuitem_move_node_smoke_test(self):
 
     @patch("django.contrib.messages.error")
     def test_menuitem_move_node_cant_move_outside_of_root(self, mocked_messages):
```
Comment (Aiky30) [inline tests/test_admin.py]: """A node can't be moved outside of a root node because that would pull it out of a tree altogether
"""

File: djangocms_navigation/admin.py [hunk 8]
```diff
@@ -241,8 +273,15 @@ def move_node(self, request, menu_content_id):
         # Disallow moving of a node on anything other than a draft version
         if self._versioning_enabled:
             menu_content = get_object_or_404(
-                MenuContent._base_manager, id=menu_content_id
+                self.MenuModel._base_manager, id=menu_content_id
             )
+            request.menu_content_id = menu_content_id
+            change_perm = self.has_change_permission(request, menu_content)
+            if not change_perm:
+                msg = 'The item is currently locked or you don\'t have permission to change it'
```
Comment (Aiky30) [inline djangocms_navigation/admin.py]: The should use the get text _("")

I think the message should also be different for version locking and when version locking is not installed e.g. 

Version locking enabled: _("The item is currently locked or you don\'t have permission to change it")
Version locking disabled: _("You don\'t have permission to change this item")

File: djangocms_navigation/admin.py [hunk 9]
```diff
@@ -102,32 +116,43 @@ def get_preview_link(self, obj):
 
 
 class MenuItemAdmin(TreeAdmin):
+    MenuModel = MenuContent
     form = MenuItemForm
     change_list_template = "admin/djangocms_navigation/menuitem/change_list.html"
     list_display = ["__str__", "get_object_url"]
 
+    def get_admin_name(self, name, reverse=False):
```
Comment (Aiky30) [inline djangocms_navigation/admin.py]: I would say that this should be called get_admin_reverse_url() as it is returning the url and not an admin name which the naming would suggest.

File: djangocms_navigation/admin.py [hunk 10]
```diff
@@ -102,32 +116,43 @@ def get_preview_link(self, obj):
 
 
 class MenuItemAdmin(TreeAdmin):
+    MenuModel = MenuContent
```
Comment (Aiky30) [inline djangocms_navigation/admin.py]: this should probably be menu_model, look at form below: form = MenuItemForm

File: djangocms_navigation/admin.py [hunk 11]
```diff
@@ -46,9 +52,22 @@ def url_for_result(self, result):
 
 class MenuContentAdmin(admin.ModelAdmin):
     form = MenuContentForm
+    ItemModel = MenuItem
     list_display = ["title", "get_menuitem_link", "get_preview_link"]
     list_display_links = None
 
+    def get_admin_name(self, name, reverse=False, model=None):
```
Comment (Aiky30) [inline djangocms_navigation/admin.py]: Why does this method have the optional model and the same method in MenuItemAdmin does not?

Comment (Aiky30) [reply djangocms_navigation/admin.py]: It would be better to create a utility I think, call it something like get_model_admin_reverse in this projects utils.py like: https://github.com/divio/django-cms/blob/develop/cms/utils/urlutils.py#L95

File: djangocms_navigation/cms_config.py [hunk 12]
```diff
@@ -83,6 +86,8 @@ def on_menu_content_archive(version):
 
 
 class NavigationCMSAppConfig(CMSAppConfig):
+    # NOTE what to do in sitemap. Very similar. It will register the correct model
+    # with the config. But navigation models may be different. Settings for this?
```
Comment (jonathan-s) [inline djangocms_navigation/cms_config.py]: @Aiky30 How would you like to solve this, any ideas? The only difference is `navigation_models`.

Comment (jonathan-s) [reply djangocms_navigation/cms_config.py]: I got this.

File: tests/test_admin.py (hunk 3 above)
Comment (jonathan-s) [reply tests/test_admin.py]: Fixed.

File: tests/test_admin.py (hunk 5 above)
Comment (jonathan-s) [reply tests/test_admin.py]: Fixed.

File: djangocms_navigation/utils.py [hunk 13]
```diff
@@ -1,15 +1,53 @@
...
+    return url
+
+
 @lru_cache(maxsize=1)
 def supported_models():
     try:
-        app_config = apps.get_app_config("djangocms_navigation")
+        MenuContent = get_model('MENU_MODEL')
+        app_label = MenuContent._meta.app_label
+        app_config = apps.get_app_config(app_label)
```
Comment (Aiky30) [inline djangocms_navigation/utils.py]: This is smart, I like this

File: djangocms_navigation/utils.py [hunk 14]
```diff
@@ -19,7 +57,9 @@ def supported_models():
 
 @lru_cache(maxsize=1)
 def supported_content_type_pks():
-    app_config = apps.get_app_config("djangocms_navigation")
+    MenuContent = get_model('MENU_MODEL')
+    app_label = MenuContent._meta.app_label
+    app_config = apps.get_app_config(app_label)
```
Comment (Aiky30) [inline djangocms_navigation/utils.py]: I'm thinking that because this is repeated this block should be a utility to get the apps config, an internal `get_app_config_from_model()` Returns app_config

File: tests/test_utils.py [hunk 15]
```diff
@@ -23,23 +23,28 @@ def setUp(self):
...
 
-    def test_supported_models(self):
-        models = ["Foo", "Bar"]
-        app_config = Mock(
-            spec=[], cms_extension=Mock(spec=[], navigation_apps_models=models)
-        )
-        with patch.object(apps, "get_app_config", return_value=app_config):
-            self.assertEqual(supported_models(), models)
+    def test_supported_models_with_installed_apps(self):
+        """Tests that all models registered in cms_config shows
```
Comment (Aiky30) [inline tests/test_utils.py]: Try not to start a test with: "Tests that", we all do it and this comment was said to me: "It's obvious that it's testing that because it's a test"

You can list the business / technical requirement or edge case:
"All models registered in cms_config are shown as supported models"

File: tests/test_utils.py [hunk 16]
```diff
@@ -23,23 +23,28 @@ def setUp(self):
...
+            PollContent: ["text"],
+        }
+        self.assertDictEqual(registered_models, expected_models)
 
     @patch.object(apps, "get_app_config", side_effect=LookupError)
     def test_supported_models_returns_empty_list_on_lookup_error(self, mocked_apps):
         self.assertDictEqual(supported_models(), {})
 
     def test_supported_models_is_cached(self):
-        models = ["Foo", "Bar"]
```
Comment (Aiky30) [inline tests/test_utils.py]: This looks liek something Monika would have done. If it was created by Monika there is a reason that she's mocking here. I won't reject this change but try and understand why she may have mocked here.

File: djangocms_navigation/constants.py [hunk 17]
```diff
@@ -21,8 +21,9 @@ def get_templates():
     return choices
 
 
-PLUGIN_URL_NAME_PREFIX = "djangocms_navigation"
-
-SELECT2_CONTENT_OBJECT_URL_NAME = "{}_select2_content_object".format(
-    PLUGIN_URL_NAME_PREFIX
-)
+def get_select2_url_name():
```
[the remark this answers is not in the document]
Comment (jonathan-s) [reply djangocms_navigation/constants.py]: Happy to move this, don't feel strongly about it.

File: tests/test_utils.py (hunk 16 above)
Comment (jonathan-s) [reply tests/test_utils.py]: Using the app_config as a mock object will get you an error. Generally I'm of the opinion if you can use real objects you should as that will help you avoid bugs. For this particular example I can't find a reason to use a mock for the app_config.  

In this case we're testing the cache and with the approach below it absolutely fulfills that criteria.

File: djangocms_navigation/admin.py [hunk 18]
```diff
@@ -292,10 +313,9 @@ def get_object_url(self, obj):
     @property
     def _versioning_enabled(self):
         """Helper property to check if versioning is enabled for navigation"""
-        return apps.get_app_config(
```
[the remark this answers is not in the document]
Comment (jonathan-s) [reply djangocms_navigation/admin.py]: The overview of this patch is that it makes djangocms-navigation re-usable. The current trade-off that's been made is that if you've got several apps that need to use djangocms-navigation AND these apps need to have customized models that will be somewhat difficult. (ie the the global setting of menu model + the above). 

If we need to support djangocms-navigation being re-usable and having multiple apps using customized djangocms-navigation models. I think I would need to spend more time working out the edge cases and write tests that explicitly tests this.  

For reference, this is what the config in sitemap would look like. 
```
class HeadlessSitemapCMSAppConfig(CMSAppConfig):
    navigation_models = {
        # model_class : field(s) to search in menu item form UI
        Page: ["title"],
        Url: ["internal_name"]
    }

    fil_headless_deployment_enabled = True
    deployment_serializer_backends = {MenuContent: SitemapSerializerBackend}
```
If you include `djangocms_versioning_enabled = True` in the headless sitemap you would also need to include. 

```
        versioning = [
            VersionableItem(
                content_model=MenuContent,
                grouper_field_name="menu",
                copy_function=copy_menu_content,
                preview_url=MenuContent.get_preview_url,
                on_publish=on_menu_content_publish,
                on_unpublish=on_menu_content_unpublish,
                on_draft_create=on_menu_content_draft_create,
                on_archive=on_menu_content_archive,
            )
        ]
```

When you do MenuContent would be registered once in the sitemap addon and once in djangocms-navigation which would throw an error if I remember correctly.

File: djangocms_navigation/utils.py (hunk 13 above)
Comment (jonathan-s) [reply djangocms_navigation/utils.py]: Thank you!

**Comment by jonathan-s** [comment]
The following issue cropped up when developing. 
https://github.com/divio/djangocms-url-manager/issues/26

File: tests/test_views.py [hunk 19]
```diff
@@ -1,12 +1,11 @@
 from django.contrib.contenttypes.models import ContentType
 from django.contrib.sites.models import Site
-from django.urls.exceptions import NoReverseMatch
 
 from cms.models import Page, User
 from cms.test_utils.testcases import CMSTestCase
 from cms.utils.urlutils import admin_reverse
 
-from djangocms_navigation.constants import get_select2_url_name
+from djangocms_navigation.utils import get_select2_url_name
```
Comment (Aiky30) [inline tests/test_views.py]: This should be stacked in alphabetical order with models, then test_utils then utils.

File: djangocms_navigation/cms_config.py [hunk 20]
```diff
@@ -1,67 +1,103 @@
...
+                    grouper_field_name="menu",
+                    copy_function=complete_copy_menu_content,
+                    preview_url=MenuContent.get_preview_url,
+                    on_publish=on_menu_content_publish,
+                    on_unpublish=on_menu_content_unpublish,
+                    on_draft_create=on_menu_content_draft_create,
+                    on_archive=on_menu_content_archive,
+                )
+            ]
+            cms_config.versioning = versioning
```
Comment (jonathan-s) [inline djangocms_navigation/cms_config.py]: Add the following code after line 61

```
            from cms.utils.setup import configure_cms_apps
            versioning_app = apps.get_app_config('djangocms_versioning')
            # will throw an error because configuring versioning apps is not idempotent
            configure_cms_apps([versioning_app])
```

File: djangocms_navigation/cms_config.py [hunk 21]
```diff
@@ -1,71 +1,73 @@
...
 from cms.app_base import CMSAppConfig, CMSAppExtension
 from cms.models import Page
 
-from .utils import purge_menu_cache, get_model
-from .constants import MENU_MODEL, ITEM_MODEL
-
-
-MenuContent = get_model(MENU_MODEL)
-MenuItem = get_model(ITEM_MODEL)
+from .utils import purge_menu_cache
```
Comment (Aiky30) [inline djangocms_navigation/cms_config.py]: tiny change, the .utils import should come after .models

File: djangocms_navigation/cms_config.py [hunk 22]
```diff
@@ -1,71 +1,73 @@
...
+        if isinstance(navigation_app_models, dict):
+            self.navigation_apps_models.update(navigation_app_models)
+        else:
+            raise ImproperlyConfigured(
+                "navigation configuration must be a dictionary object"
+            )
+
 
-def copy_menu_content(original_content):
+def copy_menu_content(menu_model, item_model, original_content):
```
Comment (Aiky30) [inline djangocms_navigation/cms_config.py]: We don't know if this could have been reused by other apps, because of this I think we should make the additional attributes optional: `def copy_menu_content(original_content, menu_model=MenuContent, item_model=MenuItem):`. Partial won't work in this scenario and we would have to wrap the function. Any ideas of how we could do that?

The solution designed here will be used as an example for any addons that reuse copy methods. I think it's important that we make them backwards compatible.

Comment (jonathan-s) [reply djangocms_navigation/cms_config.py]: You are right that it's better to make this backwards compatible. Making the signature the above looks to be the correct choice. 

However I think partial still does work. Take a look at the python docs: https://docs.python.org/3.6/library/functools.html#functools.partial

It should work with keyword arguments as well.
```
>>> from functools import partial
>>> basetwo = partial(int, base=2)
>>> basetwo.__doc__ = 'Convert base 2 string to an int.'
>>> basetwo('10010')
```

Comment (jonathan-s) [reply djangocms_navigation/cms_config.py]: You are right that it's better to make this backwards compatible. Making the signature the above looks to be the correct choice. 

However I think partial still does work. Take a look at the python docs: https://docs.python.org/3.6/library/functools.html#functools.partial

It should work with keyword arguments as well.
```
>>> from functools import partial
>>> basetwo = partial(int, base=2)
>>> basetwo.__doc__ = 'Convert base 2 string to an int.'
>>> basetwo('10010')
```

Comment (jonathan-s) [reply djangocms_navigation/cms_config.py]: Looks like I was wrong. How about we solve it this way?

```
def _copy_menu_content(original_content, menu_model, item_model):
   # snip 
    return new_content


complete_copy_menu_content = partial(
    _copy_menu_content,
    MenuContent,
    MenuItem
)


def copy_menu_content(original_content):
    """Copy the MenuContent object and deepcopy its menu items."""
    return complete_copy_menu_content(original_content)
```

Comment (jonathan-s) [reply djangocms_navigation/cms_config.py]: That way `copy_menu_content` keeps its original signature.

File: djangocms_navigation/utils.py [hunk 23]
```diff
@@ -2,9 +2,34 @@
...
+# FIXME: Reuse the django cms reverse util: admin_reverse()
+def reverse_admin_name(model, name, args=None, kwargs=None):
+    name = get_admin_name(model, name)
+    url = reverse('admin:{}'.format(name), args=args, kwargs=kwargs)
+    return url
+
+
+def get_select2_url_name():
+    url_name = "{}_select2_content_object".format(
+        MenuContent._meta.app_label
```
Comment (Aiky30) [inline djangocms_navigation/utils.py]: Make the model configurable for this method.
